broker/pkg/coordinator/loader.go

```diff
@@ -166,13 +167,52 @@ func (c *Loader) loadMessages(ctx context.Context) bool {
 	return !allFinished
 }
 
-func (c *Loader) hydrateOneConsumer(ctx context.Context, tuple *ConsumerPipelineTuple) (int, error) {
-	return 0, nil
+func (c *Loader) hydrateOneConsumer(ctx context.Context, limit int, tuple *ConsumerPipelineTuple) (int, error) {
+	fullRange := domain.BucketRange{
+		Start: 0,
+		End:   c.conf.Topic.Settings.BucketCount - 1,
+	}
+
+	var sent int
+	var err error
+	var pushLeaseMessages []timeline.Lease
+
+	defer tuple.C.AddAvailableBufferSize(-uint32(sent))
+	consumerLimit := limit
+	if int(tuple.C.LoadAvailableBufferSize()) < limit {
+		consumerLimit = int(tuple.C.LoadAvailableBufferSize())
+	}
+	hasMoreForThisBucket := true // we presume it has
+	for hasMoreForThisBucket {
+		pushLeaseMessages, hasMoreForThisBucket, err = c.storage.SelectByConsumer(ctx, tuple.C.GetTopicAsBytes(), tuple.C.GetIDAsBytes(), fullRange, consumerLimit, dtime.TimeToMS(time.Now())+c.conf.PrefetchMaxMilliseconds)
+		if err != nil {
+			return sent, err
+		}
+		if len(pushLeaseMessages) == 0 {
+			return sent, nil
+		}
+
+		for i := range pushLeaseMessages {
+			if pushLeaseMessages[i].Message.GetID() == "" {
```

Comment:
as I know we already solve the problem when some messages were returned with empty ids, I think You should not check this anymore. More than this a message that have an empty ID is not a valid message and we have to solve this issue at storage level (SelectByConsumer)